misc/memory-connection-limits/tests/util.rs

```diff
@@ -0,0 +1,119 @@
+// Copyright 2023 Protocol Labs.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::task::{Context, Poll};
+
+use libp2p_core::{Endpoint, Multiaddr};
+use libp2p_identity::PeerId;
+use libp2p_swarm::{
+    dummy, ConnectionClosed, ConnectionDenied, ConnectionId, FromSwarm, NetworkBehaviour,
+    PollParameters, THandler, THandlerInEvent, THandlerOutEvent, ToSwarm,
+};
+use void::Void;
+
+#[derive(libp2p_swarm_derive::NetworkBehaviour)]
+#[behaviour(prelude = "libp2p_swarm::derive_prelude")]
+pub(crate) struct TestBehaviour {
+    pub(crate) connection_limits: libp2p_memory_connection_limits::Behaviour,
+    pub(crate) mem: ConsumeMemoryBehaviour,
+}
+
+#[derive(Default)]
+pub(crate) struct ConsumeMemoryBehaviour {
+    map: Vec<Vec<u8>>,
+}
+
+impl ConsumeMemoryBehaviour {
+    fn handle_connection(&mut self) {
+        // 1MB
+        self.map.push(vec![1; 1024 * 1024]);
```

Comment:
Let's make the 1MB configurable and pass it in from the test. I was initially confused on why the test works because there was no visible connection!

Reply:
I've made it a const generic and use a type alias `ConsumeMemoryBehaviour1MBPending0Established` for better readability.

```rust
let mut network = Swarm::new_ephemeral(|_| TestBehaviour {
        connection_limits: Behaviour::with_max_percentage(0.1),
        mem_consumer: ConsumeMemoryBehaviour1MBPending0Established::default(),
    });
```

Reply:
That works too!